Mathlib/Logic/Equiv/Fin.lean

```diff
@@ -491,6 +492,43 @@ def finProdFinEquiv : Fin m × Fin n ≃ Fin (m * n)
 #align fin_prod_fin_equiv_apply_val finProdFinEquiv_apply_val
 #align fin_prod_fin_equiv_symm_apply finProdFinEquiv_symm_apply
 
+/-- The equivalence induced by `a ↦ (a / n, a % n)` for nonzero `n`.
+This is like `finProdFinEquiv.symm` but with `m` infinite.
+See `Nat.div_mod_unique` for a similar propositional statement. -/
+@[simps]
+def Nat.divModEquiv (n : ℕ) [NeZero n] : ℕ ≃ ℕ × Fin n where
+  toFun a := (a / n, ↑a)
+  invFun p := p.1 * n + ↑p.2
+  -- TODO: is there a canonical order of `*` and `+` here?
+  left_inv a := Nat.div_add_mod' _ _
+  right_inv p := by
+    refine' Prod.ext _ (Fin.ext <| Nat.mul_add_mod_of_lt p.2.is_lt)
+    dsimp only
+    rw [add_comm, Nat.add_mul_div_right _ _ (NeZero.pos n), Nat.div_eq_of_lt p.2.is_lt, zero_add]
+#align nat.div_mod_equiv Nat.divModEquiv
+
+/-- The equivalence induced by `a ↦ (a / n, a % n)` for nonzero `n`.
+See `Int.ediv_mod_unique` for a similar propositional statement. -/
+@[simps]
+def Int.divModEquiv (n : ℕ) [NeZero n] : ℤ ≃ ℤ × Fin n where
+  -- TODO: could cast from int directly if we import `data.zmod.defs`, though there are few lemmas
+  -- about that coercion.
+  toFun a := (a / n, ↑(a.natMod n))
+  -- porting note: need explicit `Nat.cast` else we get `Int.ofNat` as the coercion instead.
```

Comment:
[Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/simp.20normal.20form.3A.20ofNat.20or.20Nat.2Ecast.3F/near/338423289)

Reply:
#2878 will fix this.